Add methods for list publications, list contributors, post in collection + tests

mikkot opened the pull request:
Hello @majelbstoat, @dpup, 

Please review the following commits I made in branch 'mikko-nodejs-sdk-list-collections'.

8ff1da95988b761ea8ba39ebfa5c955cc4308d69 (2015-11-11 14:09:18 -0800)
Add implementation for the following methods/APIs: 
- getPublicationsForUser : GET /v1/users/:userId/publications
- getContributorsForPublication : GET /v1/publications/:publicationId/contributors
- createPostInPublication : POST /v1/publication/:publicationId/posts

Also added unit and integration tests (the latter will be a no-op unless a local config file is provided)

R=@majelbstoat
R=@dpup

mikkot commented on .gitignore:
```diff
@@ -7,6 +7,9 @@ pids
 *.seed
 
+# Configuration for integration tests
+test/config.json
```
This is not provided, which will disarm integration tests. Can be added locally to run against local (hermetic) server

majelbstoat replied:
don't think we should add this to a public repo. Can you do a local .gitignore?

mikkot replied:
What's the concern here? If people actually use said file for testing there changes, this helps them not include it in a pull request. Telling people that such file is used for testing is not really that significant - it's obvious reading the tests. If you suggest that we leave the tests out altogether I see this point, but don't necessarily agree with excluding tests. If there are solid arguments to do so happy to reconsider :)

mikkot replied:
Per Jamie's (reasonable and well argumented) request to scrap integration tests, this is gone

mikkot commented on index.js:
```diff
@@ -1,312 +1 @@
-  PostLicense: PostLicense,
-  PostContentFormat: PostContentFormat
-}
+module.exports = require("./lib/medium-client.js")
```
With adding tests, felt necessary to pull actual code into lib/, which makes this pull request harder to review. I'll add notes on medium-client.js to show where I changed things.

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+// Copright 2015 A Medium Corporation
+
+var http = require('http')
```
Added this line to make http available later

mikkot replied:
This will be gone

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+
+
+var DEFAULT_ERROR_CODE = -1
+var MISSING_REQUIRED_PARAM = 100
```
Added new error. Chose number 100 somewhat randomly

majelbstoat replied:
These numbers are not random :) come chat to me and we can discuss :)

mikkot replied:
OK

mikkot replied:
Error code will be gone

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ */
+var Scope = {
+  BASIC_PROFILE: 'basicProfile',
+  LIST_PUBLICATIONS: 'listPublications',
```
Added new scope constant

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ * @constructor
+ */
+function MediumClient(options) {
+  this._enforce(options, ['clientId', 'clientSecret'])
```
Refactored to use _enforce

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ * }}
+ * @param {NodeCallback} callback
+ */
+MediumClient.prototype.getPublicationsForUser = function (options, callback) {
```
Added this method

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ * }}
+ * @param {NodeCallback} callback
+ */
+MediumClient.prototype.getContributorsForPublication = function (options, callback) {
```
Added this method

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ * @param {NodeCallback} callback
+ */
+MediumClient.prototype.createPost = function (options, callback) {
+  this._enforce(options, ['title', 'contentFormat', 'content'])
```
Added call to _enforce to kill invalid requests before they hit the server

majelbstoat replied:
can we discuss this? generally, i prefer to have the server determine what is or is not a valid request because we can move faster that way.

mikkot replied:
That is a good point, although I think the overhead of changing the SDK with API changes is pretty minimal here. I'll catch you IRL to chat, not opposed to changing this but would like to hear the full point of view.

mikkot replied:
Changed the use of _enforce so that the code only calls it on parameters that the *client* code requires, leaving enforcement of parameters that the server needs for the server

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ * }} options
+ * @param {NodeCallback} callback
+ */
+MediumClient.prototype.createPostInPublication = function (options, callback) {
```
Added this method

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+MediumClient.prototype._makeRequest = function (options, callback) {
+  // The bypass is for testing purposes only - api.medium.com will
+  // not accept HTTP connections
+  var useSsl = !process.env.MEDIUM_API_USE_HTTP
```
Added this override for running tests locally

majelbstoat replied:
we shouldn't add this to a public SDK. this is for the whole world to use ;)

mikkot replied:
I don't see a problem here. Adding this flag does not introduce any kind of security vulnerability anywhere - anyone interested in making these requests over HTTP could change the code very easily, and this shouldn't accidentally turn HTTP on for anyone. What this gives us though is possibility of running integration tests against standard.dev:8080, which has obvious value to it.

mikkot replied:
This will be gone

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+  var useSsl = !process.env.MEDIUM_API_USE_HTTP
+  var requestParams = {
+    host: process.env.MEDIUM_API_HOST || 'api.medium.com',
+    port: useSsl ? 443 : 8080,
```
Added the override for host and used useSsl for port

mikkot replied:
... and also removed them

mikkot commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+    method: options.method,
+    path: options.path
+  }
+  var req = (useSsl ? https : http).request(requestParams, function (res) {
```
Added conditional use of http or https

mikkot replied:
Now gone

mikkot commented on package.json:
```diff
@@ -3,13 +3,15 @@
-  , "authors": [
-    "Jamie Talbot <jamie@jamietalbot.com> (https://github.com/majelbstoat)"
-  ]
+  , "author": "Jamie Talbot <jamie@jamietalbot.com> (https://github.com/majelbstoat)"
```
'author' takes a single name, there's no plural in the package.json reference. (https://docs.npmjs.com/files/package.json#people-fields-author-contributors)

mikkot commented on test/load-config.js:
```diff
@@ -0,0 +1,13 @@
+  config = JSON.parse(fs.readFileSync('./test/config.json'))
+  for (var k in config) {
+    if (config.hasOwnProperty(k)) {
+      process.env[k] = config[k];
```
This feels a bit dicey - I think it's best to pass variables to actual SDK code via env, but test code reads configs from the object returned below. Let me know what you think.

mikkot replied:
This whole file is gone now, and so is the dicey feeling

mikkot commented on test/medium-client-integration-tests.js:
```diff
@@ -0,0 +1,120 @@
+    }
+  }
+
+  describe("MediumClient", function () {
```
These unit tests are very basic and just make sure proper options are enforced for various calls. I could look into mocking http/https object and also ensuring API calls look as expected - that is only covered through integration tests for now.

mikkot replied:
Hmm .. realize this was a comment on unit tests attached to the file for integration tests. Well, integration tests are now gone and unit tests are now very thorough. Tension resolved.

majelbstoat commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+// Copright 2015 A Medium Corporation
```
filename would probably be mediumClient

mikkot replied:
Happy to change - our style guide wasn't super explicit about this (other than saying '-' is preferred over '_')

mikkot replied:
Filename changed

majelbstoat commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ *
+ * @param {{
+ *  publicationId: string
+ * }}
```
```
* @param {{
+ *  publicationId: string
+ * }} options
```

mikkot replied:
Will fix

mikkot replied:
Fixed

majelbstoat commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ *
+ * @param {{
+ *  userId: string
+ * }}
```
@param needs param name

mikkot replied:
Will fix

mikkot replied:
Fixed

majelbstoat commented on lib/medium-client.js:
```diff
@@ -0,0 +1,423 @@
+ * Creates a post on Medium and places it under specified publication.
+ *
+ * Note that following restrictions around this function:
+ * - Current user must be an editor or a writer for the specified publication
```
Not sure this kind of documentation belongs in an sdk, because we need to repeat it for every one. This seems like the kind of thing for api-docs.

mikkot replied:
Happy to remove (have another pull request for updating API docs anyway). Maybe add a link though?

mikkot replied:
Verbosity replaced with a hyperlink

majelbstoat commented on test/medium-client-unit-tests.js:
```diff
@@ -0,0 +1,115 @@
+    func.call(obj, options, function (err, data) {
+      if (err) throw err
+    })
+  }).should.throw(/^Missing required parameter/)
```
can we test an error code here, instead of a string?

mikkot replied:
Sure, can change that.

mikkot replied:
As I decided to remove the new errror code, tests are now testing or MediumError to make sure other errors potentially thrown by the code don't give false positives

majelbstoat commented:
i think we could get reasonable test coverage without any of the medium-dev specific stuff by using nock, which would allow us to assess the requests that are being made.

I'm also not convinced that we should do param checking on the client side. Just sets us up to be more brittle when server stuff changes.

mikkot commented:
I am OK improving unit test coverage with nock, that sounds like the right approach to increase coverage. If there's unease about the integration tests, I can move them to a private repo (I don't we should drop them altogether). However, I think the right solution is to have only one version of the SDK that is tested. If env variables to enable this are not the right solution, happy to hear about alternatives.

mikkot commented:
Replaced integration tests with more rigorous nock tests. Found and fixed two client bugs in the process - finding those from the PR will be left as an exercise for the reader

majelbstoat commented on lib/mediumClient.js:
```diff
@@ -0,0 +1,411 @@
+    data: qs.stringify(params)
+  }, false, function (err, data) {
+    if (!err) {
+      this._accessToken = data.access_token
```
fix number 1! :)

mikkot replied:
:+1:

majelbstoat commented on lib/mediumClient.js:
```diff
@@ -0,0 +1,411 @@
+        var err = payload.errors[0]
+        callback(new MediumError(err.message, err.code), null)
+      } else if (statusType == 2) {
+        callback(null, unwrapDataEnvelope ? payload.data : payload)
```
fix number 2! :)

Or... you could just do `callback(null, payload.data || payload)`

majelbstoat replied:
(And then avoid the unwrapDataEnvelope param)

mikkot replied:
:+1: , and I like your fix better. Let me change that

majelbstoat commented on test/mediumClient_test.js:
```diff
@@ -0,0 +1,295 @@
+        refresh_token: refreshToken
+      }
+      var request = nock('https://api.medium.com/', {
+                          'Content-Type': 'application/x-www-form-urlencoded'
```
our indentation would have this two spaces indented from `var` above.

majelbstoat commented:
One suggestion, one formatting nit, LG!

mikkot commented:
Implemented suggestion and fixed indents, tests passed (locally) -> merging this in
